signprot/views.py

```diff
@@ -511,21 +502,272 @@ def fams_tab():
                     fd[p].append('')
 
         return fd
 
+
+class CouplingBrowser2(TemplateView):
+    """
+    Class based generic view which serves coupling data between Receptors and G-proteins.
+    Data coming from Guide to Pharmacology, Asuka Inuoue and Michel Bouvier at the moment.
+    More data might come later from Roth and Strachan TRUPATH biosensor and Neville Lambert.
+    :param dataset: ProteinGProteinPair (see build/management/commands/build_g_proteins.py)
+    :return: context
+    """
+
+    template_name = "signprot/coupling_browser2.html"
+
+    @method_decorator(csrf_exempt)
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+
+        tab_fields, header = self.tab_fields()
+
+        context['tabfields'] = tab_fields
+        context['header'] = header
+        flat_list = [item for sublist in header.values() for item in sublist]
+        context['subunitheader'] = flat_list
+
+        return context
+
+    @staticmethod
+    def tab_fields():
+        """
+        This function returns the required fields for the G-protein families table and the G-protein subtypes table
+        which are to be rendered in separate tabs in the same page.
+
+        :return: key.value pairs from dictotemplate dictionary
+        keys =id values in ProteinGProteinPair table.
+        values = source, class, family, uniprotid, iupharid, logmaxec50_deg, pec50_deg, emax_deg
+        """
+
+        proteins = Protein.objects.filter(sequence_type__slug='wt',
+                                          family__slug__startswith='00',
+                                          species__common_name='Human').prefetch_related(
+            'family',
+            'family__parent__parent__parent'
+        )
+
+        couplings = ProteinGProteinPair.objects.filter(source="GuideToPharma").values_list('protein__entry_name',
+                                                                                           'g_protein__name',
+                                                                                           'transduction')
+
+        signaling_data = {}
+        for pairing in couplings:
+            if pairing[0] not in signaling_data:
+                signaling_data[pairing[0]] = {}
+            signaling_data[pairing[0]][pairing[1]] = pairing[2]
+
+
+        protein_data = {}
+        for prot in proteins:
+            protein_data[prot.id] = {}
+            protein_data[prot.id]['class'] = prot.family.parent.parent.parent.shorter()
+            protein_data[prot.id]['family'] = prot.family.parent.short()
+            protein_data[prot.id]['uniprot'] = prot.entry_short()
+            protein_data[prot.id]['iuphar'] = prot.family.name.replace('receptor', '').strip()
+            protein_data[prot.id]['accession'] = prot.accession
+            protein_data[prot.id]['entryname'] = prot.entry_name
+
+            # MAKES 2396 SQL QUERIES, have to find out how to make it faster.
```

Comment:
In order to speed this up, you can do a prefetch on the web resources while collecting the proteins at line 556:
Protein.objects.<stuff>.prefetch_related('web_links__web_resource')

Reply:
Update: this won't work because of the many-to-many relationship. If want to do reduce this to two calls: first grab the proteins, then separately grab all desired weblinks to these proteins and match them up.